src/odm_sharing/private/cons.py

```diff
@@ -23,37 +34,186 @@ def _write_table_to_db(db: sa.engine.Engine, table: str, df: pd.DataFrame
+
+    :returns: a set of column names
+    '''
+    # XXX: Strings are stored with dtype=object since they are Python objects.
+    # No other Python objects can be imported from a spreadsheet, so it should
+    # always be `str`.
+    BOOL_VALS = ['FALSE', 'TRUE']
+    NA_VALS = ['', 'NA']
+    result: Set[str] = set()
+    for col in df:
+
+        # check column type
+        if df[col].dtype == bool:
+            result.add(str(col))
+            continue
+        if df[col].dtype != object:  # not str
+            continue
+
+        # check string value: NA, bool, or other
+        for val in df[col]:
+            norm_val = val.upper()
+            if norm_val in NA_VALS:  # NA
+                continue
+            if norm_val in BOOL_VALS:  # bool
+                result.add(str(col))
+            break
+    return result
+
+
 def _connect_csv(path: str) -> Connection:
     '''copies file data to in-memory db
 
     :raises OSError:'''
+    # XXX: NA-values are not normalized to avoid mutating user data (#31)
     logging.info('importing csv file')
     table = Path(path).stem
-    db = _create_memory_db()
-    df = pd.read_csv(path)
-    _write_table_to_db(db, table, df)
-    return db
+    df = pd.read_csv(path, na_filter=False)
+    bool_cols = _get_csv_dataset_bool_cols(df)
+
+    # normalize bool (string) values to 0/1
+    #
+    # XXX: Bool columns with NA-values still represent bools as FALSE/TRUE.
+    # This needs to be normalized to 0/1, like normal bool columns, so that all
+    # bool columns have the same bool values.
+    for col in bool_cols:
+        if df[col].dtype == object:  # str
+            df[col] = df[col].replace({'FALSE': '0', 'TRUE': '1'})
+
+    db = _datasets_to_db({table: df})
+    return Connection(db, {table: bool_cols})
+
+
+def _get_excel_bool_col_ixs(data_sheet: Sheet, formula_sheet: Sheet
+                            ) -> List[int]:
+    '''returns the indices of all bool columns'''
+    # XXX: iter_rows gives a type error with mypy
+    result = []
+    header_row = next(data_sheet.iter_rows())  # type: ignore
+    column_count = len(header_row)
+    remaining = list(range(column_count))
+
+    def iter_col_cells(sheet: Sheet) -> Generator:
+        '''iterates column-indices and cells, for each row in the sheet'''
+        # XXX: the read-only workbook only has row iteration
+        # XXX: the first row is skipped since it's assumed to be the header
+        # XXX: iteration of `remaining` must use a copy since it may be mutated
+        for row in sheet.iter_rows(min_row=2):  # type: ignore
+            if len(remaining) == 0:
+                break
+            for col_ix in remaining.copy():
+                cell = row[col_ix]
+                yield (col_ix, cell)
+
+    def add_bool_col(col_ix: int) -> None:
+        result.append(col_ix)
+        remaining.remove(col_ix)
+
+    # find bool cols in data sheet
+    for col_ix, cell in iter_col_cells(data_sheet):
```

Comment:
Is there a reason you're iterating through every cell in a column as opposed to using the logic in the csv file parsing where if we encounter something that isn't an NA or bool we skip the remaining values?

Reply:
Thanks, it seems like I have to fix this part, I'll get back to you.

Reply:
I've rewritten the bool conversion in a much cleaner and correct way now.